pkg/controllers/nodeoverlay/controller.go

```diff
@@ -0,0 +1,256 @@
+	return "nodeoverlay.controller"
+}
+
+// NewController constructs a controller for node overlay validation
+func NewController(kubeClient client.Client, cp cloudprovider.CloudProvider, instanceTypeStore *InstanceTypeStore) *Controller {
+	return &Controller{
+		kubeClient:        kubeClient,
+		cloudProvider:     cp,
+		instanceTypeStore: instanceTypeStore,
+		temporaryStore:    NewInstanceTypeStore(),
+	}
+}
+
+// Reconcile validates that all node overlays don't have conflicting requirements
+func (c *Controller) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, c.Name())
+
+	overlayList := &v1alpha1.NodeOverlayList{}
+	nodePoolList := &v1.NodePoolList{}
+	if err := c.kubeClient.List(ctx, overlayList); err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodeoverlays, %w", err)
+	}
+	err := c.kubeClient.List(ctx, nodePoolList)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing nodepools, %w", err)
+	}
+
+	overlayWithRuntimeValidationFailure := map[string]error{}
+	overlaysWithConflict := []string{}
+	c.temporaryStore = NewInstanceTypeStore()
+	nodePoolToInstanceTypes := map[string][]*cloudprovider.InstanceType{}
+
+	for _, np := range nodePoolList.Items {
+		its, err := c.cloudProvider.GetInstanceTypes(ctx, &np)
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("listing instance types, %w", err)
+		}
+		nodePoolToInstanceTypes[np.Name] = its
+	}
+
+	overlayList.OrderByWeight()
+	for i := range overlayList.Items {
+		if err := overlayList.Items[i].RuntimeValidate(ctx); err != nil {
+			overlayWithRuntimeValidationFailure[overlayList.Items[i].Name] = err
+			continue
+		}
+
+		// Due to reserved capacity type offering being dynamically injected as part of the GetInstanceTypes call
+		// We will need to make sure we are validating against each nodepool to make sure. This will ensure that
+		// overlays that are targeting reserved instance offerings will be able to apply the offering.
+		for _, np := range nodePoolList.Items {
+			overlaysWithConflict = append(overlaysWithConflict, c.checkOverlayPerNodePool(np, nodePoolToInstanceTypes[np.Name], overlayList.Items[i])...)
+		}
+	}
+
+	if err := c.updateOverlayStatuses(ctx, overlayList.Items, overlaysWithConflict, overlayWithRuntimeValidationFailure); err != nil {
+		return reconcile.Result{}, fmt.Errorf("updating nodeoverlay statuses, %w", err)
+	}
+
+	c.instanceTypeStore.UpdateStore(c.temporaryStore.updates)
+	return reconcile.Result{RequeueAfter: 6 * time.Hour}, nil
+}
+
+func (c *Controller) Register(_ context.Context, m manager.Manager) error {
+	b := controllerruntime.NewControllerManagedBy(m).
+		Named(c.Name()).
+		// The reconciled overlay does not matter in this case as one reconcile loop
+		// will compare every overlay against every other overlay in the cluster.
+		For(&v1alpha1.NodeOverlay{}).
+		Watches(&v1.NodePool{}, NodeOverlayEventHandler(c.kubeClient)).
```

Comment:
I believe there's an optimization we could do here by making sure we only rereconcile on NodePools when `spec.template.spec.requirements` or `spec.template.metadata.labels` has changed. I don't believe we care about any other fields, since those are the only two fields that can impact a NodeOverlays selection criteria. We couldn't make a similar optimization for NodeClasses since we don't know which fields could impact offering / requirement generation.